### ISSUE TITLE
Proxy support

### ISSUE THREAD

**Issue by andris9** [nodemailer/nodemailer#555]
Over the years there has been several requests for adding support for proxies in Nodemailer.

  * #520 How to setting proxy?
  * #171 Request for proxy support
  * #125 Mail sending behind proxy
  * etc.

One solution would be to allow passing around already established connections. In this case the client could open a socket through proxy to destination and then pass this socket to be used for SMTP. Pooled connections require a different approach since sockets are created and killed all the time, this could be solved by providing a callback mechanism for the transporter that can be called every time a new socket is needed. Client creates a new socket and returns it to transporter.

Once a general socket handling mechanism exist, I could probably add some built-in support for known proxy types (HTTP CONNECT, SOCKS4, SOCKS5)

**Comment by andris9** [comment]
smtp-connection v2.1.0 has support for external sockets that nodemailer-smtp.* could use. The final API should go something like this:

```javascript
transport.onCreateSocket = function(callback){
     var socket = openProxyConnection(function(){
        callback(null, socket);
    });
};
```

**Comment by andris9** [comment]
Implemented in v2.2.0-beta.2
